### PR TITLE
[webui][api] interface to access a project's public key and ssl certificate

### PR DESCRIPTION
**Pull request by evanrolfe**
We want users to be able to access a project's public key and its meta data from the web ui and the api. There is a new backend api endpoint to get this information: `/source/$PROJECT/_keyinfo`.

Github issue: https://github.com/openSUSE/open-build-service/issues/2472

Who: public users
What: want to access a project's public key and its metadata
Why: in order to verify the project's signature

### CODE REVIEW

File: src/api/config/environments/development.rb [hunk 1]
```diff
@@ -8,8 +8,8 @@
   # since you don't have to restart the web server when you make code changes.
   config.cache_classes = false
 
-  # Do not eager load code on boot.
-  config.eager_load = false
+  # Eager load code on boot.
+  config.eager_load = true
```
Comment (evanrolfe) [inline src/api/config/environments/development.rb]: A note on this change: because I added a folder `app/controllers/webui/project/` it meant that any controller code within the `app/controllers/webui` would look for a module in that folder whenever they called the `Project` model. I think its because of rails autoloading not correctly resolve the constant. However enabling eager_load in development fixes this problem. The only downside is that its said to increase startup times for the rails server but I haven't noticed a difference myself.

There are some other solutions in case we dont want to enable eager_load:

1. Add `require 'project'` in the `WebuiController` - I'm not such a fan of this because it doesn't fix the problem for other developers who may add another folder to the webui controllers in the future i.e. adding `app/controllers/webui/package/` would still cause an "Unitialised constant" error and cause that developer to spend time trying to figure out the solution unless they happen to already know about this.

2. Change all calls to the `Project` model in the webui controllers to `::Project` - I think this is a viable solution but also more work..

File: src/api/app/models/public_key.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+class PublicKey
```
Comment (ChrisBr) [inline src/api/app/models/public_key.rb]: We already have ``app/models/backend_file.rb``, maybe you can inherit from it?

File: src/api/config/environments/development.rb (hunk 1 above)
Comment (ChrisBr) [inline src/api/config/environments/development.rb]: I don't mind. We already use the ``::`` several times in the app.
However, please move this in a separate commit at least as it is basically unrelated to your other changes.

File: src/api/spec/controllers/webui/project/ssl_certificate_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
+require 'rails_helper'
+require 'webmock/rspec'
+
+RSpec.describe Webui::Project::SslCertificateController, type: :controller, vcr: true do
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/project/ssl_certificate_controller_spec.rb]: As you don't add cassettes do you really need ``vcr: true``?
https://github.com/vcr/vcr

File: src/api/spec/controllers/webui/project/public_key_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
+require 'rails_helper'
+require 'webmock/rspec'
+
+RSpec.describe Webui::Project::PublicKeyController, type: :controller, vcr: true do
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/project/public_key_controller_spec.rb]: As you don't add cassettes do you really need vcr: true?
https://github.com/vcr/vcr

File: src/api/spec/controllers/project/public_key_controller_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
+require 'rails_helper'
+require 'webmock/rspec'
+
+RSpec.describe Project::PublicKeyController, type: :controller, vcr: true do
```
Comment (ChrisBr) [inline src/api/spec/controllers/project/public_key_controller_controller_spec.rb]: As you don't add cassettes do you really need vcr: true?
https://github.com/vcr/vcr

File: src/api/config/environments/development.rb (hunk 1 above)
Comment (evanrolfe) [inline src/api/config/environments/development.rb]: Thanks for the review @ChrisBr! Just to check, are you saying you don't mind having `config.eager_load = true` or you don't mind using `::Project` in the webui controllers?

Also FYI, there is a third option too:

3. Change the folder name to `app/controllers/webui/projects/` so the module doesn't conflict with the model. This is the preferred way for organising rails controllers but it wouldn't be consisted with the existing controller and route naming like `ProjectController`

File: src/api/spec/controllers/webui/project/ssl_certificate_controller_spec.rb (hunk 3 above)
Comment (evanrolfe) [inline src/api/spec/controllers/webui/project/ssl_certificate_controller_spec.rb]: Ah good spot. I was testing out vcr here earlier thats why its still there. Will remove these..

File: src/api/config/environments/development.rb (hunk 1 above)
Comment (Ana06) [inline src/api/config/environments/development.rb]: What about pluralizing the project controller? then the plural folder would be consistent. :thinking:

File: src/api/spec/controllers/project/public_key_controller_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    let(:backend_url) { CONFIG['source_url'] + PublicKey.send(:backend_url, project.name) }
+
+    before do
+      stub_request(:get, backend_url).and_return(body: keyinfo_response)
+
+      get :show, params: { format: :xml, project_name: project.name }
+    end
+
+    context 'with a project that has a public key and an ssl certificate' do
+      let(:gpg_public_key) { Faker::Lorem.characters(1024) }
```
Comment (Ana06) [inline src/api/spec/controllers/project/public_key_controller_controller_spec.rb]: This line is repeated, you could put it up and write it only once :wink:

File: src/api/spec/controllers/webui/project/public_key_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  describe 'GET #show' do
+    let(:project) do
```
Comment (Ana06) [inline src/api/spec/controllers/webui/project/public_key_controller_spec.rb]: why not writing it like

``` Ruby
let(:project) { create(:project, name: "test_project", title: "Test Project") }
```

Too many lines :confounded:

File: src/api/spec/controllers/webui/project/ssl_certificate_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  describe 'GET #show' do
+    let(:project) do
```
Comment (Ana06) [inline src/api/spec/controllers/webui/project/ssl_certificate_controller_spec.rb]: Too many lines :confounded:

File: src/api/spec/controllers/webui/project/ssl_certificate_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    let(:backend_url) { CONFIG['source_url'] + PublicKey.send(:backend_url, project.name) }
+
+    before do
+      stub_request(:get, backend_url).and_return(body: keyinfo_response)
+
+      get :show, params: { project_name: project.name }
+    end
+
+    context 'with a project that has an ssl certificate' do
+      let(:gpg_public_key) { Faker::Lorem.characters(1024) }
```
Comment (Ana06) [inline src/api/spec/controllers/webui/project/ssl_certificate_controller_spec.rb]: Repeated line!

File: src/api/app/models/public_key.rb (hunk 2 above)
Comment (ChrisBr) [inline src/api/app/models/public_key.rb]: ^^

File: src/api/config/environments/development.rb (hunk 1 above)
Comment (evanrolfe) [inline src/api/config/environments/development.rb]: @Ana06 see my last post about pluralising the project controller. Though I think thats my favorite solution too.. @ChrisBr what do you think? Do you still prefer to use `::Project`?

**Comment by evanrolfe** [comment]
[(3) Interface to access a project's public key](https://trello.com/c/WDDjOPfX/159-3-interface-to-access-a-project-s-public-key)

File: src/api/app/controllers/application_controller.rb [hunk 10]
```diff
@@ -569,6 +569,17 @@ def set_response_format_to_xml
     request.format = :xml if request.format == :html
   end
 
+  def extract_user_public
```
Comment (mdeniz) [inline src/api/app/controllers/application_controller.rb]: Isn't better if this method lives only in the Project::PublicKeyController ??, is only used there. 
Or you plan to use it in another controller?.

BTW, I know that you only moved the code, but... can be better like that?
```ruby
def extract_user_public
  if ::Configuration.anonymous
    load_nobody
  else
    logger.error 'No public access is configured'
    render_error( message: 'No public access is configured', status: 401 )
  end
end
```

File: src/api/app/controllers/project/public_key_controller.rb [hunk 11]
```diff
@@ -0,0 +1,17 @@
+class Project
```
Comment (mdeniz) [inline src/api/app/controllers/project/public_key_controller.rb]: Is this a class or a module?

File: src/api/app/models/public_key.rb (hunk 2 above)
Comment (mdeniz) [inline src/api/app/models/public_key.rb]: Agree 😄

**Comment by evanrolfe** [comment]
@mdeniz I think the reason why so many vcr cassettes are changed is because this new feature effects the [user's home page](https://github.com/evanrolfe/open-build-service/blob/bf8e4b794c5cde8f90879bec8f8c8204b8230c30/src/api/app/views/webui/project/show.html.erb#L144) (the page the user is redirected to after login / sign up) so that it always makes a request to `localhost:5352/source/#{project_name}/_keyinfo?withsslcert=1` to check if the project has a public key. 

This means that any feature spec which logs in a user (most of them) will need to have that backend url stubbed by vcr. Hence why so many cassettes are changed.

Does that make sense?

File: src/api/app/controllers/project/public_key_controller.rb (hunk 11 above)
Comment (evanrolfe) [inline src/api/app/controllers/project/public_key_controller.rb]: In this particular case it is a class because we already have the `Project` model which is a class. Though I will be changing this to `module Projects` (plural) anyway so this will become a module..

Comment (mdeniz) [inline src/api/app/controllers/project/public_key_controller.rb]: nice 👍

File: src/api/app/controllers/application_controller.rb (hunk 10 above)
Comment (evanrolfe) [inline src/api/app/controllers/application_controller.rb]: @mdeniz yeah I like that style much better too, but doesn't it still need to include the `return true`? Like this?
```ruby
def extract_user_public
  if ::Configuration.anonymous
    load_nobody
    return true
  else
    logger.error 'No public access is configured'
    render_error( message: 'No public access is configured', status: 401 )
  end
end
```

Comment (evanrolfe) [inline src/api/app/controllers/application_controller.rb]: Actually you're correct. This method is only ever used in a `before_action` so the returns are pointless..

Comment (mdeniz) [inline src/api/app/controllers/application_controller.rb]: 👍

File: src/api/app/controllers/projects/public_key_controller.rb [hunk 12]
```diff
@@ -0,0 +1,17 @@
+module Projects
```
Comment (hennevogel) [inline src/api/app/controllers/projects/public_key_controller.rb]: I would like to discuss controller modularity more broadly before introducing this into the codebase...

**Comment by adrianschroeter** [comment]
Please make _keyinfo also accessable via the generic api route, not only /public (or do I missed that?)

**Comment by evanrolfe** [comment]
Sorry I'm having some problems now with this PR where travis keeps trying to checkout the branch from opensuse's github when it only exist on my own github account. I think its happening because I accidentally pushed this branch to opensuse but then deleted it when I realised what happened.

I'm closing this PR and opening a new one here with the same code:
[https://github.com/openSUSE/open-build-service/pull/2544](https://github.com/openSUSE/open-build-service/pull/2544)

File: src/api/app/controllers/application_controller.rb (hunk 10 above)
Comment (evanrolfe) [inline src/api/app/controllers/application_controller.rb]: @mdeniz this method was already being using by `PublicController` hence why its in `ApplicationController`..

https://github.com/evanrolfe/open-build-service/blob/feature%2Fpublic_key_info_interface2/src/api/app/controllers/public_controller.rb#L21
